core/src/main/scala/org/http4s/EntityDecoder.scala

```diff
@@ -106,11 +106,19 @@ object EntityDecoder extends EntityDecoderInstances {
       msg.headers.get(`Content-Type`) match {
         case Some(contentType) =>
           if (a.matchesMediaType(contentType.mediaType)) a.decode(msg, strict)
-          else b.decode(msg, strict)
+          else b.decode(msg, strict).leftMap{
+            case MediaTypeMismatch(actual, expected) =>
+              MediaTypeMismatch(actual, expected ++ a.consumes)
+            case other => other
+          }
 
         case None =>
           if (a.matchesMediaType(UndefinedMediaType)) a.decode(msg, strict)
-          else b.decode(msg, strict)
+          else b.decode(msg, strict).leftMap{
+            case MediaTypeMissing(expected) =>
+              MediaTypeMissing(expected ++ a.consumes)
```

Comment:
We don't restrict the exception type on the left, so I'm not sure who would be making that expectation.  Also note that a MediaTypeMissing could come through with an *invalid* Content-Type header (a fact I learned a few minutes ago trying out the XSS issue).  But I agree that `UndefinedMediaType` value does look a little strange vs. having a method.